### PR TITLE
[TRTLLM-4987][feat] Support generation logits in TRTLLMSampler

### PR DESCRIPTION
**Pull request by amitz-nv**
## Description

This PR adds generation logits support in TRTLLMSampler:

- An algorithm `handle_generation_logits` is created in Python that mimics the C++ counterpart, currently without speculative decoding support, and beam width > 1 support was not tested and may not work properly, as it is outside the scope of this task.
- Adjusted `test_return_logits.py` to also validate the argmax of the returned generation logits are as expected, also testing it with & without overlap scheduler.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 1]
```diff
@@ -75,21 +78,36 @@ def append(self, logits: torch.Tensor):
...
+        except IndexError:
+            return None
 
-    def get(self, all_logits=False):
-        start = 0 if all_logits else self.last_position
-        return self._storage[start:self.
-                             position] if self._storage is not None else None
+    def get_all(self):
+        try:
+            last_index = -2 if self._should_exclude_last else -1
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:103]: I assume there is a special case when the length of the container is 1, please cover that here.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 2]
```diff
@@ -75,21 +78,36 @@ def append(self, logits: torch.Tensor):
...
+            max_position = self._logits_indices[last_index][1]
+            return self._storage[:max_position]
+        except IndexError:
+            return None
```
Comment (dcampora) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:107]: I suggest to replace `get_at` and `get_all` with the following succinct implementation:

```
def get(self, all_logits=False):
    try:
        last = -2 if self._should_exclude_last else -1
        start = 0 if all_logits else self._logits_indices[last][0]
        end = self._logits_indices[last][1]
        return self._storage[start:end]
    except IndexError:
        return None
```

**Review by dcampora — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 3]
```diff
@@ -37,11 +37,14 @@
 
 class LogitsStorage:
 
-    def __init__(self, seq_length: int, use_device_memory=True):
+    def __init__(self,
+                 seq_length: int,
+                 use_device_memory=True,
+                 should_exclude_last=False):
```
[the remark this answers is not in the document]
Comment (amitz-nv) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:43]: I agree it is preferred, but I'm not sure the sampler type is known to every location that creates `LlmRequest`, for example, how can I know the sampler type in [`executor_request_to_llm_request` function in `tensorrt_llm/_torch/pyexecutor/llm_request.py` ](https://github.com/NVIDIA/TensorRT-LLM/blob/6c144b5a2f4600706b3829e5b86b08dfda80f256/tensorrt_llm/_torch/pyexecutor/llm_request.py#L357)? I hope I'm just missing something, as I definitely agree setting it in the ctor is preferred

[the remark this answers is not in the document]
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:43]: That happens in the `py_executor` instance, which should know about the sampler used. You can check the type of `self.sampler` for this.

[the remark this answers is not in the document]
Comment (amitz-nv) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:43]: Done and also changed `isinstance(self.sampler, TRTLLMSampler)` to `not isinstance(self.sampler, TorchSampler)`, so when `TorchSampler` is removed, that condition couldn't be accidentally missed.
